### PR TITLE
async for authorization

### PR DESCRIPTION
**Pull request by Geobert**
I'm opening the PR to get help because it doesn't compile and I'm kinda stuck :( 

```
error[E0499]: cannot borrow `authorization` as mutable more than once at a time
   --> oxide-auth\src\code_grant\authorization.rs:332:27
    |
332 |         requested = match authorization.advance(input) {
    |                           ^^^^^^^^^^^^^ mutable borrow starts here in previous iteration of loop
```

I'm wondering why I get this error when I'm doing things in a similar way to other flows :-/

### CODE REVIEW

**Comment by HeroicKatora** [comment]
I can't do a full review right now, however if you use to state machine style of `accesstoken` and don't borrow the request but rather clone the state you need you should be able to avoid most of the borrow checker intricacies. Don't think too much about overhead right now, most often such a clone happens anyways in flow already and the most expensive part should be smaller allocations which we can address later by allowing reuse of it in the flows.

**Comment by Geobert** [comment]
Thanks! I'll give that a shot :)

**Comment by Geobert** [comment]
Hum, I tried to clone everything instead of borrowing, tried to extract stuff that I used to store in the state machine to create the `Pending` (outside the `loop` of the driver) and still getting this error.

What do you mean when you say that I'm using `refresh` style? Because I'm comparing my code with `access_token` one and I can't see the difference :-/

**Comment by Geobert** [comment]
Got an answer, my `Requested` had a lifetime, which break it all. I don't really understand why though

**Comment by HeroicKatora** [comment]
That's more or less what I meant: get rid of lifetime parameters on _all_ structs and instead own and clone what you need. I see now that `accesstoken` is not yet based on the state machine in its free standing method, so `resource` would have been the better blue print. `refresh` borrows from the request, and its state has a lifetime parameter—`resource`'s state does not have a lifetime and does not borrow from the request. The same is true for the [`Requested` in its `protect` method](https://github.com/HeroicKatora/oxide-auth/blob/async/oxide-auth/src/code_grant/resource.rs#L213-L214).

The cause of your borrow issue is two-fold: Firstly, you use the same lifetime parameter `'req` for both the input and the state, thus when you borrow one the other must be treated as borrowed as well but also requires your input to live and be borrowed for the same time as the state. And secondly, the `Output` which borrows from the statemachine is fed into the loops `Requested` which keeps the statemachine borrowed as well. You should either use the required parameters directly or clone those portions which you will use in the next loop iteration.

**Comment by Geobert** [comment]
I've pushed the state machine, but don't merge yet, I'm working on the async part of this now.

I've change the free function of accesstoken to use the state machine as well :)

**Comment by Geobert** [comment]
I think this time it's ok :)

File: oxide-auth/src/code_grant/authorization.rs [hunk 1]
```diff
@@ -79,93 +277,135 @@ pub trait Endpoint {
...
+        request.state().clone(),
+        AuthorizationError::default(),
+    );
+    err.description().set_type(err_type);
+    err
 }
 
 /// Represents a valid, currently pending authorization request not bound to an owner. The frontend
 /// can signal a reponse using this object.
+#[derive(Clone)]
```
[the remark this answers is not in the document]
Comment (Geobert) [reply oxide-auth/src/code_grant/authorization.rs:408]: Good catch, I don't even remember why I put that in the first place as the compiler doesn't complain about it being removed.

File: oxide-auth/src/code_grant/authorization.rs [hunk 2]
```diff
@@ -69,6 +70,203 @@ pub trait Endpoint {
     fn extension(&mut self) -> &mut dyn Extension;
 }
 
+/// The result will indicate wether the authorization succeed or not.
+pub struct Authorization<'req> {
+    state: AuthorizationState,
+    bound_client: Option<BoundClient<'req>>,
```
[the remark this answers is not in the document]
Comment (Geobert) [reply oxide-auth/src/code_grant/authorization.rs:76]: The thing is that bound_client is not needed in the state that follows the one where it's produced: It's given as an input in `Input::Bound` but then we have the state `AuthorizationState::Extending` which doesn't need it and then after, we have `AuthorizationState::Negotiating` which needs it. 

What's best then? Carrying `bound_client` in Extending without using it or storing it as I'm doing?

**Pull request merged**